### ISSUE TITLE
RStudio shows wrong variable labels in the data viewer.

### ISSUE THREAD

**Issue by iseppo** [rstudio/rstudio#9771]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.4.1717, 1.4.1725
    OS Version      : Linux
    R Version       : 4.1.1

### Steps to reproduce the problem

```r
df <- data.frame(a=1, b=2, c=3)
attr(df, "variable.labels") <- c(a = "a", b = "b", c = "c")
View(df) #variable labels showed correctly
library(dplyr)
View(select(df,  -b)) #wrong variable labels showed
```

### Describe the problem in detail
Rstudio shows wrong variable labels. It shows variable labels by their position in attributes(df)$variable.labels, not by matching them by variables. So it shows here "b" as variable c label, not "c" as expected.

### Describe the behavior you expected

Rstudio shows correct variable labels.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@iseppo Thank you for raising this! I'll mark it for review as part of our ongoing development of RStudio.

Reproducible in
- RStudio Desktop 1.4.1717 on MacOS 12 beta
- RStudio Server 1.4.1717 on Ubuntu 20.04 via Safari on MacOS 12 beta
- RStudio Server 2021.09.0-preview+338 on Ubuntu 20.04 via Safari on MacOS 12 beta

**Comment by iseppo** [comment]
As some background - this is something that seems to have changed recently (probably something changing in the way R deals with variable labels, keeping the old variable labels, exposing this bug in RStudio viewer, but I haven't been able to find the specific change).

It affects some workloads a lot. Specifically people working with questionnaire data, which usually come in as SPSS files and are read in with variable labels, which people rely on instead of the variable names. The moment they remove some variables they can end up making major mistakes.

**Comment by GCRev** [comment]
The `select` function doesn't modify the list of labels, so the R function that re-associates the labels doesn't know any better than to display "b" where it should be "c." 

I think we need someone who has more expertise with R to take a look at either the `select` function or the `src/cpp/session/modules/SessionDataViewer.R` file in order to see what a fix for this might look like. I wouldn't classify this as a UI issue, but I'm happy to learn more about the backend in order to be able to fix issues like this in the future.

**Comment by kevinushey** [comment]
```
> s <- select(uhu, -b)
> str(s)
'data.frame':	1 obs. of  2 variables:
 $ a: num 1
 $ c: num 3
 - attr(*, "variable.labels")= Named chr [1:3] "a" "b" "c"
  ..- attr(*, "names")= chr [1:3] "a" "b" "c"
```

You can see that the `dplyr` select statement did not subset `variable.labels` as you expect. This appears to be a `dplyr` issue to me.

**Comment by iseppo** [comment]
> The `select` function doesn't modify the list of labels, so the R function that re-associates the labels doesn't know any better than to display "b" where it should be "c."

Its a named vector, so even if dplyr's behaviour is unexpected it can be fixed in RStudio by finding which label corresponds to which variable. We know it, it is just not that the second label corresponds to second variable, but we have to find the variable label corresponding to the variable name.

**Comment by iseppo** [comment]
The issue would be fixed with 
changing the line 135 in src/cpp/session/modules/SessionDataViewer.R
from:
         col_label <- colLabels[[idx]]
to:
         col_label <- colLabels[[col_name]]
I do not know if there would be any usecases that would break though.

**Comment by kevinushey** [comment]
@iseppo: do you want to file a pull request?

Note that your change would need to handle the case where `variable.labels` is unset or, for whatever reason, is not actually a named character vector.

**Comment by iseppo** [comment]
I have created a pull request at https://github.com/rstudio/rstudio/pull/10082.
It is my first time, so I am sorry if I missed something.

**Comment by romainfrancois** [comment]
I opened https://github.com/tidyverse/dplyr/issues/6136 if we want to deal with this on the `dplyr` end.

**Comment by jgutman** [comment]
Merged into 2022.02.0 on January 10, should have been closed.
